vpdq/cpp/hashing/filehasher.cpp

```diff
@@ -62,74 +92,285 @@ static void saveFrameToFile(AVFrame* frame, const char* filename) {
-    AVCodecContext* codecContext,
-    std::unique_ptr<vpdq::hashing::AbstractFrameBufferHasher>& phasher,
-    vector<hashing::vpdqFeature>& pdqHashes,
-    double frameRate,
-    bool verbose,
-    int frameNumber,
-    int frameMod) {
-  // Send the packet to the decoder
-  int ret = avcodec_send_packet(codecContext, packet) < 0;
-  if (ret < 0) {
-    throw std::runtime_error("Cannot send packet to decoder");
+static AVFramePtr createTargetFrame(int width, int height) {
+  // Create a frame for resizing and converting the decoded frame to RGB24
+  AVFramePtr frame(av_frame_alloc());
+  if (frame.get() == nullptr) {
+    throw std::bad_alloc();
   }
 
-  // Receive the decoded frame
-  while (ret >= 0) {
-    ret = avcodec_receive_frame(codecContext, frame);
-    if (ret == AVERROR(EAGAIN) || ret == AVERROR_EOF) {
-      break;
-    } else if (ret < 0) {
-      throw std::runtime_error("Cannot receive frame from decoder");
+  frame->format = PIXEL_FORMAT;
+  frame->width = width;
+  frame->height = height;
+
+  if (av_image_alloc(
+          frame->data, frame->linesize, width, height, PIXEL_FORMAT, 1) < 0) {
+    throw std::bad_alloc();
+  }
+  return frame;
+}
+
+class AVVideo {
+ public:
+  AVCodecContext* codecContext = nullptr;
+  SwsContext* swsContext = nullptr;
+  AVFormatContext* formatContext = nullptr;
+  int videoStreamIndex = -1;
+  int width;
+  int height;
+  double frameRate;
+
+  AVVideo(const std::string filename) {
+    // Open the input file
+    if (avformat_open_input(
+            &formatContext, filename.c_str(), nullptr, nullptr) != 0) {
+      throw std::runtime_error("Cannot open video");
+    }
+
+    // Retrieve stream information
+    if (avformat_find_stream_info(formatContext, nullptr) < 0) {
+      avformat_close_input(&formatContext);
+      throw std::runtime_error("Cannot find video stream info");
     }
 
-    if (frameNumber % frameMod == 0) {
-      // Resize the frame and convert to RGB24
-      sws_scale(
-          swsContext,
-          frame->data,
-          frame->linesize,
-          0,
-          codecContext->height,
-          targetFrame->data,
-          targetFrame->linesize);
-      // Call pdqHasher to hash the frame
-      int quality;
-      pdq::hashing::Hash256 pdqHash;
-      bool ret = phasher->hashFrame(targetFrame->data[0], pdqHash, quality);
-      if (!ret) {
-        throw std::runtime_error(
-            "Failed to hash frame buffer." + std::string("Frame: ") +
-            std::to_string(frameNumber) +
-            std::string(
-                " Frame width or height smaller than the minimum hashable dimension"));
+    // Find the first video stream
+    for (unsigned int i = 0; i < formatContext->nb_streams; ++i) {
+      if (formatContext->streams[i]->codecpar->codec_type ==
+          AVMEDIA_TYPE_VIDEO) {
+        videoStreamIndex = i;
+        break;
       }
+    }
+
+    if (videoStreamIndex == -1) {
+      avformat_close_input(&formatContext);
```

Comment:
Hmm, I wonder if separating construction and opening will let you lean on your destructor for correctness, otherwise you have to carefully partially undo everything allocated.

Another solution might be to wrap the pointers with another object with a destructor, and if the constructor fails it can still destruct the inner object normally.  

Something I read while looking at this: https://isocpp.org/wiki/faq/exceptions#selfcleaning-members

Reply:
I could move constructor to an init, but I have a hunch that might not be a good idea. An AVVideo instance shouldn't exist without belonging to an opened video.

It might just be more readable, although verbose, to just free everything manually. The open video sequence is very straightforward.

If you move the complexity outside of the class with init, then that responsibility falls on them to handle it. 

Wrapping the pointers would be a must if they were used at all outside of the AVVideo class other than reads, but since they're not right now I think it would just add more complexity than it would solve.

Your call though, either way makes sense.

Reply:
I think this might not be too much work:

## Option 1: Swap all pointers to smartpointers with custom deallocs
Before:
```
AVCodecContext* codecContext = nullptr;
SwsContext* swsContext = nullptr;
AVFormatContext* formatContext = nullptr;
```
After:
```
# lots of custom desctructors
unique_ptr<AVCodecContext> codecContext = nullptr;
unique_ptr<SwsContext> swsContext = nullptr;
unique_ptr<AVFormatContext> formatContext = nullptr;
```

## Option 2: Inner object to manage pointers:
class AVContext {
  AVCodecContext* codecContext = nullptr;
  SwsContext* swsContext = nullptr;
  AVFormatContext* formatContext = nullptr;

  open() {
    codecContext = alloc(...) || throw()
    swsContext = alloc(...) || throw()
    formatContext = alloc(...) || throw()
  }

  ~Foo(){
    if (codecContext != nullptr) {
      dealloc(...)
    }
   if (swsContext != nullptr) {
      dealloc(...)
    }
    ...
  }
}

class AVVideo {
 public:
  AVContext context{}; // Might be forgetting the exact syntax

  AVVideo() {
    // any exceptions will lead to dealloc called on context, so don't have to manage order
    context.open();
  }